### PR TITLE
Specification for storage system detection

### PR DESCRIPTION
**Pull request by shtripat**
Signed-off-by: Shubhendu <shtripat@redhat.com>

### CODE REVIEW

**Comment by shtripat** [comment]
@Tendrl/tendrl-core @Tendrl/tendrl_api please review

**Comment by mkudlej** [comment]
Do I understand it right that after implementing this spec there is no need to add sds_version(and sds_name) parameter to API function calls?
You can find sds_version and sds_name as parameters here https://github.com/anivargi/documentation/blob/c26ccba15091efd70de41c54929f5f710626b8ec/api/ceph-pool-examples.adoc and https://github.com/Tendrl/documentation/blob/master/api/gluster-create-volume-example.adoc

If yes, then there is API impact and it should be mentioned in spec.

**Comment by shtripat** [comment]
@mkudlej These values get tagged with storage nodes. I feel still jobs need to provide sds_type and sds_version fields fetched from nodes list while import cluster flows.

@brainfunked comments?

**Comment by mkudlej** [comment]
@shtripat My question was not just about cluster import but more about all other functions for managing clusters and other cluster atoms.

File: specs/sds_detection_modules.adoc [hunk 1]
```diff
@@ -0,0 +1,260 @@
...
+None
+
+==== Tendrl/common impact:
+
+* Common module should define a generic interface for plugins which would
+be extended by different auto-detection modules specific interfaces. The
+structure of the generic interface could be as below
+
+```
+class AutoDetectPlugin(object):
```
Comment (cybernth) [inline specs/sds_detection_modules.adoc]: why this is required?

File: specs/sds_detection_modules.adoc [hunk 2]
```diff
@@ -0,0 +1,260 @@
...
+* Tag the dectected details (if any) with the storage nodes details
+
+* Enhance the import cluster flows in node agent component to decide on what
+Tendrl integration module gets deployed on the underlying storage nodes
+
+=== Alternatives
+
+None
+
+=== Data model impact:
```
Comment (cybernth) [inline specs/sds_detection_modules.adoc:56]: Data model and tendrl defintions needs to be added

File: specs/sds_detection_modules.adoc [hunk 3]
```diff
@@ -0,0 +1,260 @@
...
+
+* Define a standard path from where all the available storage detection modules
+would be loaded
+
+* While starting the node agent service on the storage nodes, execute these
+plugins and auto detect the storage systems (if any) on underlying storage nodes
+
+* Tag the dectected details (if any) with the storage nodes details
+
+* Enhance the import cluster flows in node agent component to decide on what
```
Comment (cybernth) [inline specs/sds_detection_modules.adoc]: I think this is out of scope as far as this spec is concer?

File: specs/sds_detection_modules.adoc [hunk 4]
```diff
@@ -0,0 +1,260 @@
...
+
+=== Data model impact:
+
+None
+
+=== Impacted Modules:
+
+==== Tendrl API impact:
+
+* API layer should attach the storage system specific auto-detected details of
```
Comment (cybernth) [inline specs/sds_detection_modules.adoc]: are you planning to do this as part of this spec. i feel limit the scope this spec only to detection of cluster details and updating the etcd

File: specs/sds_detection_modules.adoc [hunk 5]
```diff
@@ -0,0 +1,260 @@
...
+
+* SDS Detect plugin for `gluster`
+
+** Execute the command `gluster --version` to figure out if gluster bits are
+installed on the storage node. Also get the version details and tag with the
+node in central store
+
+** Default assign the role as `glusterd` for all the storage nodes
+
+** Execute the command `gluster pool list` and calculate as hash out of node
```
Comment (cybernth) [inline specs/sds_detection_modules.adoc]: Does this changes if the cluster is expanded after detecting the cluster?

File: specs/sds_detection_modules.adoc [hunk 6]
```diff
@@ -0,0 +1,260 @@
...
+UUIDs of the nodes in the trusted storage pool. Tag this as cluster_id for the
+nodes. Storage the details in pre-defined location (say
```
Comment (cybernth) [inline specs/sds_detection_modules.adoc:259]: I think https://github.com/Tendrl/specifications/pull/117 talks about detected_clusterand this information is supposed go there

**Review by cybernth — CHANGES_REQUESTED** [review]

File: specs/sds_detection_modules.adoc (hunk 3 above)
Comment (shtripat) [reply specs/sds_detection_modules.adoc]: Ok. will remove this.

File: specs/sds_detection_modules.adoc (hunk 2 above)
Comment (shtripat) [reply specs/sds_detection_modules.adoc:56]: Done. will add how node entity would be tagged with additional data.

File: specs/sds_detection_modules.adoc (hunk 4 above)
Comment (shtripat) [reply specs/sds_detection_modules.adoc]: Sure. Will remove from this spec.

File: specs/sds_detection_modules.adoc (hunk 1 above)
Comment (shtripat) [reply specs/sds_detection_modules.adoc]: This is the base interface defined in commons module (with basic functions to initialize etc). The node agent would define SDSDetectPlugin interface extended from this (with abstract function detect_storage_system). This SDSDetectPlugin would be implemented by Ceph and Gluster specific classes which would work as detectors for underlying storage system.

I feel this is the same hierarchy we discussed as well earlier. Do you think something else?

File: specs/sds_detection_modules.adoc (hunk 5 above)
Comment (shtripat) [reply specs/sds_detection_modules.adoc]: This would be used to create a unique UUID to identify the current cluster state before import. Once cluster imported into Tendrl, it could be kept as is and no need to change the cluster_id value based on newer/fewer nodes in the cluster at later stage.

@brainfunked  comments?

File: specs/sds_detection_modules.adoc (hunk 6 above)
Comment (shtripat) [reply specs/sds_detection_modules.adoc:259]: Line 117 talk about CephSDSDetectionPlugin so I didnt add gluster specific details there. Here under *SDS Detect plugin for gluster* I have added detailed commands etc to be executed.

File: specs/sds_detection_modules.adoc (hunk 1 above)
Comment (cybernth) [reply specs/sds_detection_modules.adoc]: I don't think we need a top level base class in common, only pluginmount is required here. Also Rohan pointed to a doc where some frameworks could help here. Good to refer that as well. --> http://docs.openstack.org/developer/stevedore/

Comment (shtripat) [reply specs/sds_detection_modules.adoc]: ok

Comment (shtripat) [reply specs/sds_detection_modules.adoc]: @r0h4n If we plan to use http://docs.openstack.org/developer/stevedore/ do we need to build some package from openstack? How this dependency would be brought in Tendrl?

For sure it looks like serving the purpose we need by using `extension.ExtensionManager` and we would be able to load of the extension available as plugins for detection.

File: specs/sds_detection_modules.adoc [hunk 7]
```diff
@@ -49,47 +49,104 @@ plugins and auto detect the storage systems (if any) on underlying storage nodes
...
+          type: Create
+          uuid: eda0b13a-7362-48d5-b5ca-4b6d6533a5ab
+      attrs:
+        cmd_str:
+          type: String
+        fqdn:
+          type: String
+        status:
+          type: Boolean
+        cluster_id:
```
Comment (cybernth) [inline specs/sds_detection_modules.adoc]: I think this is not the right way. Don't add as attribute, rather do it as different object like node_context

**Review by cybernth — CHANGES_REQUESTED** [review]

File: specs/sds_detection_modules.adoc (hunk 7 above)
Comment (shtripat) [reply specs/sds_detection_modules.adoc]: I am ok with this but my understanding was that these details need to be tagged with node inventory.
@brainfunked @r0h4n @nnDarshan comments?

File: specs/sds_detection_modules.adoc (hunk 1 above)
Comment (shtripat) [reply specs/sds_detection_modules.adoc]: As discussed we would go ahead with existing logic for loading the plugins dynamically and in later releases would try to use `stevedore` module for this.

File: specs/sds_detection_modules.adoc (hunk 7 above)
Comment (shtripat) [reply specs/sds_detection_modules.adoc]: Changed the logic to tag the additional details as part of Node_context

File: specs/sds_detection_modules.adoc [hunk 8]
```diff
@@ -0,0 +1,301 @@
...
+
+```
+class NodeContext(EtcdObj):
+    __name__ = 'nodes/%s/Node_context'
+
+    node_id = fields.StrField("node_id")
+    machine_id = fields.StrField("machine_id")
+    fqdn = fields.StrField("fqdn")
+    cluster_id = fields.StrField("cluster_id")
+    sds_pkg_name = fields.StrField("sds_pkg_name")
```
Comment (brainfunked) [inline specs/sds_detection_modules.adoc:68]: Drop both sds_pkg_name and sds_pkg_version. Use only roles. Put up a global namespace for /roles. Each detection module has a hardcoded role that it assigns, eg, ceph/mon, gluster/glusterd etc. The global hierarchy should contain a mapping of roles to all the nodes that belong to that role. Each node itself simply lists the roles in the roles array.

When querying, this would enable the global listing to be queries for all the nodes that are running a ceph mon or glusterd etc. This can be cross referenced against a specific cluster's list of of nodes or against the node listing provided in an import cluster call to verify that all the nodes are being detected.

Also, the sds_pkg_name and sds_pkg_version don't belong. The detection should not be based on rpm packages. It should be based on processes, services, sockets or well known file locations, such as the cli. Versions must be queried by contacting the storage system directly, using the CLI or API call if known. In any case, the objects for the detected cluster are the proper place to put the package information if required, it doesn't belong in the node context.

File: specs/sds_detection_modules.adoc [hunk 9]
```diff
@@ -0,0 +1,301 @@
...
+
+* The node agent component would introduce two python modules for ceph and
+gluster. These would be specific classes which extend from above SDSDetectPlugin
+and provide the actual implementation for the function
+
+```
+class DetectCephStorageSystem(SDSDetectPlugin):
+    def detect_storage_system(self):
+        # Implement the logic to get the details of underlying storage system
+        # 1. Figure out if ceph bits installed and set the type accordingly
```
Comment (brainfunked) [inline specs/sds_detection_modules.adoc:166]: As commented above, binaries, files, sockets, running processes etc. must be looked for, not just packages.

File: specs/sds_detection_modules.adoc [hunk 10]
```diff
@@ -0,0 +1,301 @@
...
+        # 2. Figure the version of ceph bits installed and accordingly set the
```
Comment (brainfunked) [inline specs/sds_detection_modules.adoc:167]: Using the storage system's official facility to do so ie CLI or API. Package versions have no bearing on the actual version of the storage system and are not the de-facto source of this information. Storage system's interfaces must be used to determine this information.

File: specs/sds_detection_modules.adoc [hunk 11]
```diff
@@ -0,0 +1,301 @@
...
+        # version
+        # 3. Figure out the role of the storage node (by figuring out if MON or
```
Comment (brainfunked) [inline specs/sds_detection_modules.adoc:169]: Wrong. One detection module detects only one specific role. Which means that there would be a separate module for ceph mon and ceph osd. Only the mon module would query for the cluster map to get the peer list, osd module would not. Each of these modules assigns it's own specific tag if the detection is successful.

File: specs/sds_detection_modules.adoc [hunk 12]
```diff
@@ -0,0 +1,301 @@
...
+execute these plugins one by one to figure out the details of the underlying
+storage system. Tag the figured out details with the storage nodes in central
+store
+
+* SDS Detect plugin for `ceph`
+
+** Execute the command `ceph version` to figure out if ceph bits are installed
+and figure out the version details well
+
+** Check the running status of mon and osd processes on the storage nodes and
```
Comment (brainfunked) [inline specs/sds_detection_modules.adoc:242]: One specific type of process per module only.

File: specs/sds_detection_modules.adoc [hunk 13]
```diff
@@ -0,0 +1,301 @@
...
+decide the role of the storage node accordingly
+
+** Get the FSID of the underlying cluster and tag the same as cluster_id. Store
```
Comment (brainfunked) [inline specs/sds_detection_modules.adoc:245]: This is the detected cluster id. Store it to clearly indicate as such. It is required only till a cluster is imported.

File: specs/sds_detection_modules.adoc [hunk 14]
```diff
@@ -0,0 +1,301 @@
...
+
+* SDS Detect plugin for `gluster`
+
+** Execute the command `gluster --version` to figure out if gluster bits are
+installed on the storage node. Also get the version details and tag with the
+node in central store
+
+** Default assign the role as `glusterd` for all the storage nodes
+
+** Execute the command `gluster pool list` and calculate a hash out of node
```
Comment (brainfunked) [inline specs/sds_detection_modules.adoc:257]: pool list?

**Review by brainfunked — CHANGES_REQUESTED** [review]

File: specs/sds_detection_modules.adoc (hunk 14 above)
Comment (shtripat) [reply specs/sds_detection_modules.adoc:257]: `gluster pool list` returns all the nodes in the cluster with uuids and was suggested by Atin.

File: specs/sds_detection_modules.adoc (hunk 13 above)
Comment (shtripat) [reply specs/sds_detection_modules.adoc:245]: Will change the names a `detected_cluster_id` rather

File: specs/sds_detection_modules.adoc (hunk 12 above)
Comment (shtripat) [reply specs/sds_detection_modules.adoc:242]: Done

File: specs/sds_detection_modules.adoc (hunk 10 above)
Comment (shtripat) [reply specs/sds_detection_modules.adoc:167]: We would be using `ceph version` or `gluster --version` to get the details which is CLI. May be the wordings need to be changed a bit :)

File: specs/sds_detection_modules.adoc (hunk 14 above)
Comment (japplewhite) [inline specs/sds_detection_modules.adoc:257]: Is there any negative checking here against a list of known supported configs: OS, Gluster and Ceph version, etc. Also the installation stories have lots of things we should check for like ntp, proper hostname setup, etc. Is this the appropriate place for that kind of thing?
